### ISSUE TITLE
Traefik V2 (alpha8) : TLS options not applied ?

### ISSUE THREAD

**Issue by pbenefice** [containous/traefik#5114]
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [x] No

### What did you do?

I set up a simple whoami service and try to configure specific TLS Options to the associated router

> I'm not sure this is a bug or if I'm missing something obvious, but I really tried to follow the documentation and it does not seems to work, so I'm filling this issue.
I can't open a topic on https://community.containo.us/ for now as my account is temporary on hold.

### What did you expect to see?

I expect to see no errors and get my TLS Option properly applied to my router

### What did you see instead?

I get **unknown TLS options: mintls13@file** on the *whoami* router.

I got this file provider log: 

```text
level=debug msg="Configuration received from provider file: {\"http\":{},\"tcp\":{},\"tls\":{}}" providerName=file
```

It seems my TLS config for options and defaultcert are not acknowledged by traefik as the tls section is empty :  
```
{
  "http": {},
  "tcp": {},
  "tls": {}
}
```

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.0.0-alpha8
Codename:     faisselle
Go version:   go1.12.6
Built:        2019-07-01T17:46:52Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

Complete example with all files and config to reproduce available [here](https://github.com/pbenefice/wip_traefik_tls)  

docker-compose  
```
version: "3.3"

services:

  traefik:
    image: "traefik:v2.0.0-alpha8-alpine"
    command: "--configFile=/etc/traefik/config/static/traefik.toml"
    container_name: "traefik"
    ports:
      - "0.0.0.0:80:80"
      - "0.0.0.0:443:443"
      - "127.0.0.1:8080:8080"
    volumes:
      - "./config:/etc/traefik/config"
      - "./secrets/pki/clients:/etc/pki/clients:ro"
      - "./secrets/pki/ca/ca.pem:/etc/pki/ca/ca.pem:ro"
      - "/var/run/docker.sock:/var/run/docker.sock:ro"

  whoami:
    image: containous/whoami
    expose:
      - "80"
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.whoami.entrypoints=https"
      - "traefik.http.routers.whoami.tls=true"
      - "traefik.http.routers.whoami.rule=Host(`calvin.benefice.io`)"
      - "traefik.http.routers.whoami.tls.options=mintls13@file"
```

static config  
```toml
[Global]
	CheckNewVersion = false
	SendAnonymousUsage = true

[Log]
  Level = "DEBUG"
  FilePath = ""
  Format = "common"

[API]
  Dashboard = true

[Providers]
  ProvidersThrottleDuration = 2

  [Providers.Docker]
    Watch = true
    ExposedByDefault = false
    endpoint = "unix:///var/run/docker.sock"
    SwarmMode = false

  [Providers.File]
    Watch = true
    Directory = "/etc/traefik/config/dynamic"

[entryPoints]

   [entryPoints.http]
      address = ":80"

   [entryPoints.https]
      address = ":443"
```

dynamic config  
```toml
[tls]

  [[tls.certificates]]
    certFile = "/etc/pki/clients/client.pem"
    keyFile = "/etc/pki/clients/client-key.pem"
    stores = ["default"]

  [tls.stores]
    [tls.stores.default]
      [tls.stores.default.defaultCertificate]
        certFile = "/etc/pki/clients/client.pem"
        keyFile = "/etc/pki/clients/client-key.pem"

  [tls.options]

    [tls.options.mintls13]
      minVersion = "VersionTLS13"

    [tls.options.mutualauth]
      [tls.options.mutualauth.clientAuth]
        caFiles = ["/etc/pki/ca/ca.cert"]
        clientAuthType = "RequireAndVerifyClientCert"
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
traefik    | time="2019-07-17T10:01:10Z" level=info msg="Configuration loaded from file: /etc/traefik/config/static/traefik.toml"
traefik    | time="2019-07-17T10:01:10Z" level=info msg="Traefik version 2.0.0-alpha8 built on 2019-07-01T17:46:52Z"
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Static configuration loaded {\"global\":{\"sendAnonymousUsage\":true},\"serversTransport\":{\"maxIdleConnsPerHost\":200},\"entryPoints\":{\"http\":{\"address\":\":80\",\"transport\":{\"lifeCycle\":{\"graceTimeOut\":10000000000},\"respondingTimeouts\":{\"idleTimeout\":180000000000}},\"forwardedHeaders\":{}},\"https\":{\"address\":\":443\",\"transport\":{\"lifeCycle\":{\"graceTimeOut\":10000000000},\"respondingTimeouts\":{\"idleTimeout\":180000000000}},\"forwardedHeaders\":{}},\"traefik\":{\"address\":\":8080\",\"transport\":{\"lifeCycle\":{\"graceTimeOut\":10000000000},\"respondingTimeouts\":{\"idleTimeout\":180000000000}},\"forwardedHeaders\":{}}},\"providers\":{\"providersThrottleDuration\":2000000000,\"docker\":{\"watch\":true,\"endpoint\":\"unix:///var/run/docker.sock\",\"defaultRule\":\"Host(`{{ normalize .Name }}`)\",\"swarmModeRefreshSeconds\":15000000000},\"file\":{\"directory\":\"/etc/traefik/config/dynamic\",\"watch\":true,\"traefikFile\":\"/etc/traefik/config/static/traefik.toml\"}},\"api\":{\"entryPoint\":\"traefik\",\"dashboard\":true},\"log\":{\"level\":\"DEBUG\",\"format\":\"common\"}}"
traefik    | time="2019-07-17T10:01:10Z" level=info msg="\nStats collection is enabled.\nMany thanks for contributing to Traefik's improvement by allowing us to receive anonymous information from your configuration.\nHelp us improve Traefik by leaving this feature on :)\nMore details on: https://docs.traefik.io/basics/#collected-data\n"
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="No default certificate, generate one"
traefik    | time="2019-07-17T10:01:10Z" level=info msg="Starting provider aggregator.ProviderAggregator {}"
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Start TCP Server" entryPointName=http
traefik    | time="2019-07-17T10:01:10Z" level=info msg="Starting provider *file.Provider {\"directory\":\"/etc/traefik/config/dynamic\",\"watch\":true,\"traefikFile\":\"/etc/traefik/config/static/traefik.toml\"}"
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Start TCP Server" entryPointName=https
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Start TCP Server" entryPointName=traefik
traefik    | time="2019-07-17T10:01:10Z" level=info msg="Starting provider *docker.Provider {\"watch\":true,\"endpoint\":\"unix:///var/run/docker.sock\",\"defaultRule\":\"Host(`{{ normalize .Name }}`)\",\"swarmModeRefreshSeconds\":15000000000}"
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Configuration received from provider file: {\"http\":{\"middlewares\":{\"redirect-to-https\":{\"redirectScheme\":{\"scheme\":\"https\"}}}},\"tcp\":{},\"tls\":{}}" providerName=file
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Adding certificate for domain(s) calvin.benefice.io"
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="No default certificate, generate one"
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Provider connection established with docker 18.09.2 (API 1.39)" providerName=docker
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Configuration received from provider docker: {\"http\":{\"routers\":{\"http-catchall\":{\"entryPoints\":[\"http\"],\"middlewares\":[\"redirect-to-https@file\"],\"service\":\"traefik_traefik\",\"rule\":\"hostregexp(`{host:[a-z-.]+}`)\"},\"whoami\":{\"entryPoints\":[\"https\"],\"service\":\"whoami_traefik\",\"rule\":\"Host(`calvin.benefice.io`)\",\"tls\":{\"options\":\"mintls13\"}}},\"services\":{\"traefik_traefik\":{\"loadBalancer\":{\"servers\":[{\"url\":\"http://192.168.48.2:80\"}],\"passHostHeader\":true}},\"whoami_traefik\":{\"loadBalancer\":{\"servers\":[{\"url\":\"http://192.168.48.3:80\"}],\"passHostHeader\":true}}}},\"tcp\":{}}" providerName=docker
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Adding certificate for domain(s) calvin.benefice.io"
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="No default certificate, generate one"
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Creating middleware" routerName=http-catchall@docker serviceName=traefik_traefik entryPointName=http middlewareType=Pipelining middlewareName=pipelining
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Creating load-balancer" routerName=http-catchall@docker serviceName=traefik_traefik entryPointName=http
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Creating server 0 http://192.168.48.2:80" serverName=0 entryPointName=http routerName=http-catchall@docker serviceName=traefik_traefik
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Added outgoing tracing middleware traefik_traefik" entryPointName=http routerName=http-catchall@docker middlewareName=tracing middlewareType=TracingForwarder
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Creating middleware" middlewareName=redirect-to-https@file middlewareType=RedirectScheme entryPointName=http routerName=http-catchall@docker
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Setting up redirection to https " middlewareType=RedirectScheme entryPointName=http routerName=http-catchall@docker middlewareName=redirect-to-https@file
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Adding tracing to middleware" entryPointName=http routerName=http-catchall@docker middlewareName=redirect-to-https@file
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Creating middleware" entryPointName=http middlewareName=traefik-internal-recovery middlewareType=Recovery
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Creating middleware" serviceName=whoami_traefik middlewareName=pipelining middlewareType=Pipelining entryPointName=https routerName=whoami@docker
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Creating load-balancer" serviceName=whoami_traefik entryPointName=https routerName=whoami@docker
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Creating server 0 http://192.168.48.3:80" routerName=whoami@docker serviceName=whoami_traefik serverName=0 entryPointName=https
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Added outgoing tracing middleware whoami_traefik" entryPointName=https routerName=whoami@docker middlewareName=tracing middlewareType=TracingForwarder
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="Creating middleware" entryPointName=https middlewareName=traefik-internal-recovery middlewareType=Recovery
traefik    | time="2019-07-17T10:01:10Z" level=debug msg="unknown TLS options: mintls13@docker" entryPointName=https routerName=whoami@docker
```

**Comment by dduportal** [comment]
Not 100% sure but: shouldn't the label of the `whoamitcp` container be 

```text
- "traefik.http.routers.whoami.tls.options.whoami-mintls13.minVersion=VersionTLS13"
```

?

I might be wrong, but the label should define a "TLS option" confgiuration structure, instead of referencing one from the global dynamic configuration right?

N.B. your account should have been unlocked in the Community forum. The Discourse bot locked it because less than 3s where measured between 2 messages/actions. Not sure why, sorry for the inconvenience.

**Comment by pbenefice** [comment]
Hi @dduportal, Thanks for the quick reply =)

If I refer to this section of the documentation [Here](https://docs.traefik.io/v2.0/routing/routers/#options), within the expendable block **Configuring the TLS options** I see : 
```
[http.routers]
  [http.routers.Router-1]
    rule = "Host(`foo-domain`) && Path(`/foo-path/`)"
    service = "service-id"
    # will terminate the TLS request
    [http.routers.Router-1.tls]
      options = "foo"

[tls.options]
  [tls.options.foo]
    minVersion = "VersionTLS12"
    cipherSuites = [
      "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
      "TLS_RSA_WITH_AES_256_GCM_SHA384"
    ]
```
Where it seems you need to define an Option named foo that you refer to whith `options = "foo"` in the http config.

In any case I gave it a try using the label you provided but I get the following error :  
`time="2019-07-17T14:39:58Z" level=error msg="invalid node options: string" providerName=docker container=whoami_traefik-28e0ef210aba9f1df375c86a2550492bbc1f3a75bb36ff0d36ea1d4e5b11c51e`

**Comment by dduportal** [comment]
Hi @pbenefice , sorry for the bad comment: I was wrong :)

The label `traefik.http.routers.whoami.tls.options=mintls13` is (almost) right.
It has to reference the `tls.options` structure.

But based on the last line of the traefik log you provided, it should be namespaced to the "file" provider, where the `tls13` tls option is defied, as the labels are namespaced in the `@docker` provider.

Can you confirm that it works using `traefik.http.routers.whoami.tls.options=mintls13@file` ?

**Comment by pbenefice** [comment]
Hi again @dduportal,

My mistake I forgot the @file when opening this issue as I tried multiple things before opening it, but this is something I am aware of, and I have the same issue when using it : `time="2019-07-17T15:25:55Z" level=debug msg="unknown TLS options: mintls13@file" entryPointName=https routerName=whoami@docker`

> I'll update my examples

I think the first problem is that the tls section of the configuration gathered from the file provider is empty in the logs : 
```
level=debug msg="Configuration received from provider file: {\"http\":{},\"tcp\":{},\"tls\":{}}" providerName=file
```

I might be wrong but if i receive **tls{}** as config from the provider file, I won't be able to go further.

what do you think ?

**Comment by pbenefice** [comment]
Hi again @dduportal,

For what it's worth, it seems I have a similar issue with the `tls.stores.default.defaultCertificate` directive that does not seem to be applied either.

If you have the time please feel free to use this example that should expose the problem easily : https://github.com/pbenefice/wip_traefik_tls

I'll also open a topic on the community forum.

I'm pretty sure the problem come from the fact that I receive an empty config from the file provider, but again I might miss something.

In any case thanks for your time.

**Comment by ldez** [comment]
I propose to continue the discussion on Discourse:

https://community.containo.us/t/treafik-v2-alpha8-tls-configuration-options-default-cert-not-applied/921

**Comment by pbenefice** [comment]
ok with that, so I close this issue.

**Comment by pbenefice** [comment]
Thank you @dduportal, @ldez, I manage to fix my issue by changing the provider file configuration. I used **filename** instead of **directory** in the *[Providers.File]* section, and the problem disappeared.

**Comment by dduportal** [comment]
For later readers (as the rest of the conversation took place in the Community forum):

thanks to @pbenefice , a bug in `v2.0.0-alpha8` was undercovered and is fixed by #5118. 

(Reference: https://community.containo.us/t/traefik-v2-alpha8-tls-configuration-options-default-cert-not-applied/921/7?u=dduportal)
